bin/metrics/gather.py

```diff
@@ -0,0 +1,167 @@
+# License.zenoss under the directory where your Zenoss product is installed.
+#
+##############################################################################
+import os
+import logging
+import traceback
+import json
+import time
+import argparse
+
+import requests
+
+log = logging.getLogger('zenoss.servicemetrics')
+logging.basicConfig()
+log.setLevel(logging.INFO)
+
+
+class ServiceMetrics:
+    """
+    Simple process that creates a metric gatherer, loops calling for
+    internal metrics, then posts those metrics to a consumer.
+    """
+    DEFAULT_CONSUMER = "http://localhost:22350/api/metrics/store"
+
+    def __init__(self, options):
+        if not options.service:
+            raise Exception("no service specified for which to gather metrics!")
+        self.interval = options.interval
+        self.service = options.service
+        self.metric_destination = os.environ.get("CONTROLPLANE_CONSUMER_URL", "")
+        if self.metric_destination == "":
+            self.metric_destination = self.DEFAULT_CONSUMER
+        self.session = None
+
+    def run(self):
+        gatherer = self.build_gatherer()
+        while True:
+            time.sleep(self.interval)
+            try:
+                metrics = gatherer.get_metrics()
+                self.push(metrics)
+            except Exception:
+                log.warning("Failed to gather metrics: " + traceback.format_exc())
+
+
+    def build_gatherer(self):
+        """
+        Loads up an object that can gather metrics.
+        :return: an instance of an object that implements get_metrics()
+        """
+        if self.service == "rabbitmq":
+            return RabbitMetricGatherer()
+
+    def push(self, metrics):
+        if not self.session:
+            self.session = requests.Session()
+        self.session.headers.update({'Content-Type': 'application/json'})
+        self.session.headers.update({'User-Agent': 'Zenoss Service Metrics'})
+        post_data = {'metrics': metrics}
+        response = self.session.post(self.metric_destination, data=json.dumps(post_data))
+        if response.status_code != 200:
+            log.warning("Problem submitting metrics: %d, %s" % response.status_code, response.text)
+            self.session = None
+        else:
+            log.debug("%d Metrics posted" % len(metrics))
+
+
+class MetricGatherer:
+    def __init__(self):
+        self.service_id = os.environ.get("CONTROLPLANE_SERVICED_ID", "")
+        self.instance_id = os.environ.get("CONTROLPLANE_INSTANCE_ID", "")
+        self.host_id = os.environ.get("CONTROLPLANE_HOST_ID", "")
+        self.tenant_id = os.environ.get("CONTROLPLANE_TENANT_ID", "")
+        self.tags = {
+            "serviceId": self.service_id,
+            "instance": self.instance_id,
+            "hostId": self.host_id,
+            "tenantId": self.tenant_id
+        }
+
+    def build_metric(self, name, value, timestamp, tags=None):
+        _value = self.float_value(value)
```

Comment:
Inspired by this code I originally meant to re-use: https://github.com/zenoss/zenoss-prodbin/blob/develop/Products/ZenHub/metricpublisher/utils.py#L88